# 10차시 실습 — 모델 평가와 새 데이터 예측 (정답)

`합격여부`는 1~8차시와 같은 표기입니다(1 = 합격, -1 = 불합격).

## 1단계. 9차시 모델 다시 만들기

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

df = pd.read_csv("../../data/weekly/week09/week09_pass_fail_train.csv", encoding="utf-8-sig").dropna()
print(df.shape)

(735, 9)


In [2]:
feature_cols = ["온도_섭씨", "압력_Pa", "가스유량_slm", "두께_nm", "진공도_mTorr", "습도_pct"]
X = df[feature_cols]   # 단서(Feature) 6개
y = df["합격여부"]       # 정답(Label): 1 = 합격, -1 = 불합격

In [3]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
print(len(X_train), len(X_test))

588 147


In [4]:
model = DecisionTreeClassifier(max_depth=4, random_state=42)
model.fit(X_train, y_train)

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

DecisionTreeClassifier(max_depth=4, random_state=42)

## 2단계. 모델 평가하기 — 정확도와 기준선

In [5]:
from sklearn.metrics import accuracy_score

pred = model.predict(X_test)
accuracy = accuracy_score(y_test, pred)
print(f"모델 정확도: {accuracy:.1%}")

모델 정확도: 80.3%


### 무조건 합격이라고 찍으면?

In [6]:
print(y_test.value_counts())

baseline = (y_test == 1).mean()
print(f"기준선(무조건 합격이라고 찍기) 정확도: {baseline:.1%}")

합격여부
 1    116
-1     31
Name: count, dtype: int64
기준선(무조건 합격이라고 찍기) 정확도: 78.9%


### 혼동행렬로 놓친 불합격 세기

In [7]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, pred, labels=[1, -1])
print(cm)

[[109   7]
 [ 22   9]]


## 3단계. 새 로트 불러오기

In [8]:
new_lots = pd.read_csv("../../data/weekly/week09/week09_new_lots_to_predict.csv", encoding="utf-8-sig")
print(new_lots.shape)
print(new_lots)

(9, 8)
  공정명   설비번호  온도_섭씨   압력_Pa  가스유량_slm   두께_nm  진공도_mTorr  습도_pct
0  식각  EQ-02  294.5  1005.1     50.04   96.46       4.02    44.4
1  산화  EQ-02  301.5  1007.2     55.68  100.65       5.38    46.8
2  식각  EQ-04  289.8  1009.2     50.69   96.19       5.14    42.0
3  포토  EQ-04  295.6  1019.5     49.10   98.97       4.96    44.6
4  식각  EQ-04  299.1  1026.4     56.25  103.10       4.86    38.8
5  포토  EQ-03  302.0  1026.1     51.59  103.76       4.81    50.9
6  산화  EQ-01  305.2  1005.6     47.16  101.69       4.35    42.7
7  산화  EQ-03  299.1   989.2     54.12   98.41       5.09    49.3
8  세정  EQ-01  300.6  1031.7     49.89  102.78       4.54    41.2


### 학습 때와 같은 단서만 고르기

In [9]:
X_new = new_lots[feature_cols]
print(X_new.shape)

# model.predict(new_lots)  # 주석을 풀면 ValueError: 학습 때 없던 열(공정명, 설비번호)이 섞여 있다

(9, 6)


## 4단계. 새 로트 예측하기 — predict()

In [10]:
new_lots["예측_합격여부"] = model.predict(X_new)
print(new_lots[["공정명", "설비번호", "온도_섭씨", "진공도_mTorr", "예측_합격여부"]])

  공정명   설비번호  온도_섭씨  진공도_mTorr  예측_합격여부
0  식각  EQ-02  294.5       4.02        1
1  산화  EQ-02  301.5       5.38        1
2  식각  EQ-04  289.8       5.14       -1
3  포토  EQ-04  295.6       4.96        1
4  식각  EQ-04  299.1       4.86        1
5  포토  EQ-03  302.0       4.81        1
6  산화  EQ-01  305.2       4.35        1
7  산화  EQ-03  299.1       5.09        1
8  세정  EQ-01  300.6       4.54        1


## 5단계. 확률까지 확인하기 — predict_proba()

In [11]:
proba = model.predict_proba(X_new)
print(model.classes_)
print(proba.round(3))

[-1  1]
[[0.375 0.625]
 [0.076 0.924]
 [0.614 0.386]
 [0.076 0.924]
 [0.076 0.924]
 [0.076 0.924]
 [0.375 0.625]
 [0.076 0.924]
 [0.076 0.924]]


In [12]:
new_lots["불합격_확률"] = proba[:, 0].round(3)  # classes_ [-1 1]의 0번 열 = 불합격(-1)
print(new_lots[["공정명", "설비번호", "온도_섭씨", "진공도_mTorr", "예측_합격여부", "불합격_확률"]])

  공정명   설비번호  온도_섭씨  진공도_mTorr  예측_합격여부  불합격_확률
0  식각  EQ-02  294.5       4.02        1   0.375
1  산화  EQ-02  301.5       5.38        1   0.076
2  식각  EQ-04  289.8       5.14       -1   0.614
3  포토  EQ-04  295.6       4.96        1   0.076
4  식각  EQ-04  299.1       4.86        1   0.076
5  포토  EQ-03  302.0       4.81        1   0.076
6  산화  EQ-01  305.2       4.35        1   0.375
7  산화  EQ-03  299.1       5.09        1   0.076
8  세정  EQ-01  300.6       4.54        1   0.076


### 도전 문제 — 더 깊은 나무로 다시 예측하기 (빠른 학습자용)

In [13]:
model2 = DecisionTreeClassifier(max_depth=8, random_state=42)
model2.fit(X_train, y_train)

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

DecisionTreeClassifier(max_depth=8, random_state=42)

In [14]:
print(model2.predict(X_new))

[-1  1  1  1  1  1  1  1  1]


In [15]:
print(f"깊이 4 — 학습 데이터 정확도 {model.score(X_train, y_train):.1%}, test 정확도 {model.score(X_test, y_test):.1%}")
print(f"깊이 8 — 학습 데이터 정확도 {model2.score(X_train, y_train):.1%}, test 정확도 {model2.score(X_test, y_test):.1%}")

깊이 4 — 학습 데이터 정확도 84.7%, test 정확도 80.3%
깊이 8 — 학습 데이터 정확도 94.2%, test 정확도 78.9%


**도전 문제 정리(예시)**: 깊이 8 모델은 학습 데이터 정확도가 84.7%에서 94.2%로 올랐지만 test 정확도는 80.3%에서 78.9%로 내려가 찍기(기준선)와 같아졌다. 연습문제 답은 더 잘 외웠지만 처음 보는 문제는 오히려 더 틀린 셈이라, 복잡한 모델이 항상 더 좋은 것은 아니다. 새 로트 예측도 0번·2번에서 두 모델의 답이 엇갈렸다.

## 6단계. 과정 전체 회고
학생마다 답이 다릅니다. 아래는 예시입니다.

**오늘 예측 결과(예시)**: 새 로트 9건 중 2번만 불합격(-1)으로 예측됐다(불합격 확률 61.4%). 합격(1)으로 예측된 8건 중에서도 진공도가 가장 낮은 0번·6번은 불합격 확률이 37.5%로 나머지(7.6%)보다 높았다. 정답이 없는 새 데이터라 맞았는지는 실제 검사 결과가 나와야 알 수 있다.

**10차시 큰 줄기(예시)**: 데이터를 잘 살피고, 여러 열을 함께 보며 관계(원인 후보)를 찾고, 그 관계로 예측까지 해 봤다. 다만 예측이 맞아도 원인을 안 것은 아니다.

## 7단계. AI에게 질문하며 더 알아보기
질문 예시는 강의 "AI로 한 걸음 더" 절에 있습니다. 답은 학생마다 다릅니다.

## 8단계. AI에게 코드 생성 요청하고 직접 실행해보기
강의의 "AI가 만들어줄 수 있는 코드 예시"입니다.

In [16]:
def risk_level(p):
    if p >= 0.5:
        return "높음"
    elif p >= 0.3:
        return "보통"
    else:
        return "낮음"

for p in new_lots["불합격_확률"]:
    print(f"불합격 확률 {p:.1%} → 위험도 {risk_level(p)}")

불합격 확률 37.5% → 위험도 보통
불합격 확률 7.6% → 위험도 낮음
불합격 확률 61.4% → 위험도 높음
불합격 확률 7.6% → 위험도 낮음
불합격 확률 7.6% → 위험도 낮음
불합격 확률 7.6% → 위험도 낮음
불합격 확률 37.5% → 위험도 보통
불합격 확률 7.6% → 위험도 낮음
불합격 확률 7.6% → 위험도 낮음


## 9단계. 연습문제 — 위험 로트 개수 세기

In [17]:
risky_lots = new_lots[new_lots["불합격_확률"] > 0.5]
print(f"위험 로트(불합격 확률 50% 초과) 개수: {len(risky_lots)}개")

위험 로트(불합격 확률 50% 초과) 개수: 1개


### 기준을 0.3으로 낮춰 보기

In [18]:
watch_lots = new_lots[new_lots["불합격_확률"] >= 0.3]
print(f"주의 로트(불합격 확률 30% 이상) 개수: {len(watch_lots)}개")
print(watch_lots[["공정명", "설비번호", "진공도_mTorr", "불합격_확률"]])

주의 로트(불합격 확률 30% 이상) 개수: 3개
  공정명   설비번호  진공도_mTorr  불합격_확률
0  식각  EQ-02       4.02   0.375
2  식각  EQ-04       5.14   0.614
6  산화  EQ-01       4.35   0.375


## 10단계. AI로 재미있는 미니 프로그램 만들기 — 수료증 출력하기
강의 예시 코드입니다. 자동 실행을 위해 `input()` 대신 예시 이름을 넣었습니다.

In [19]:
name = "지민"  # 실제로는 input("이름을 입력하세요: ")

print("*" * 40)
print("  반도체 데이터 분석 10차시 과정 수료증")
print(f"  이름: {name}")
print("  데이터를 살피고, 함께 보고, 예측까지 해 보았습니다.")
print("*" * 40)

****************************************
  반도체 데이터 분석 10차시 과정 수료증
  이름: 지민
  데이터를 살피고, 함께 보고, 예측까지 해 보았습니다.
****************************************
